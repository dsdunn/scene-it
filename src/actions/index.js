export const isLoading = (bool) => ({
  type: 'IS_LOADING',
  isLoading: bool
});

export const hasErrored = (bool = false) => ({
  type: 'HAS_ERRORED',
  hasErrored: bool
});

export const eventsFetchSuccess = (events) => ({
  type: 'EVENTS_FETCH_SUCCESS',
  events: events
});

export const locationFetchSuccess = (location) => ({
  type: 'LOCATION_FETCH_SUCCESS',
  location: location
});

export const selectEvent = (event) => ({
  type: "SELECT_EVENT",
  event: event
})

export const unselectEvent = () => ({
  type: "UNSELECT_EVENT"
})

