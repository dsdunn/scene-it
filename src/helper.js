import {apiKey, mapKey} from './apiKey';

export const urlBuilder = ({location, keywords}) => {
  const words = keywords.length ? `keywords=${keywords} || live+music || comedy&within=20` : 'keywords=live+music || comedy';
  const corsAnywhereUrl = 'https://cors-anywhere.herokuapp.com/'

  return `${corsAnywhereUrl}http://api.eventful.com/json/events/search?app_key=${apiKey}&${words}&location=${location.lat}, ${location.lng}&within=20&date=This Week&page_size=20`;
};

export const dataCleaner = (events) => {
  return events.map(event => ({
    cityName: event.city_name,
    eventId: event.id,
    lat: parseFloat(event.latitude),
    lng: parseFloat(event.longitude),
    performers: event.performers,
    postalCode: event.postal_code,
    region: event.region_name,
    startTime: event.start_time,
    title: event.title,
    description: event.description,
    eventUrl: event.url,
    address: event.venue_address,
    venueId: event.venue_id,
    venueName: event.venue_name,
    venueUrl: event.venue_url,
    image: event.image
  }))
}

export const reverseGeocode = async (location) => {
  const url = `https://maps.googleapis.com/maps/api/geocode/json?address=${location}&key=${mapKey}`
  const response = await fetch(url);
  const result = await response.json();
  const locality = result.results[0].address_components.find(component => component.types.includes("locality"));
  return locality.short_name;
}