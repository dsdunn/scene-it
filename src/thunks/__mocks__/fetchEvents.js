export const fetchEvents = jest.fn()