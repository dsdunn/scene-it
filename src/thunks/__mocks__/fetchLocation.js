export const fetchLocation = jest.fn();