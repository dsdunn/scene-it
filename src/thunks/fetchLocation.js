import { hasErrored, locationFetchSuccess } from '../actions';
import { mapKey } from '../apiKey';

export const fetchLocation = (location) => {
  const url = `https://maps.googleapis.com/maps/api/geocode/json?address=${location}&key=${mapKey}`
  return async (dispatch) => {
    try{
      const response = await fetch(url);
      if (!response.ok) {
        throw Error(response.statusText);
      }
      const result = await response.json();
      const coordinates = result.results[0].geometry.location;
      dispatch(locationFetchSuccess(coordinates));
    } catch (error) {
      dispatch(hasErrored(true));
    }
  };
};