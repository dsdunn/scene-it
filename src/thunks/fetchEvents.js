import { isLoading, hasErrored, eventsFetchSuccess } from '../actions';
import { dataCleaner } from '../helper';

export const fetchEvents = (url) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true));
      const response = await fetch(url,{mode:"cors"});
      if (!response.ok) {
        throw Error(response.statusText);
      }
      dispatch(isLoading(false));
      const result = await response.json();
      const cleanData = dataCleaner(result.events.event);
      dispatch(eventsFetchSuccess(cleanData));
      console.log(result.events.event);
    } catch (error) {
      dispatch(hasErrored(true));
    }
  };
};