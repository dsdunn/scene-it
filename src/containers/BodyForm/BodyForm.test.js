import React from 'react';
import { shallow, mount } from 'enzyme';
import { BodyForm } from './index';
import { mapStateToProps, mapDispatchToProps } from './index';
import { locationFetchSuccess } from '../../actions';
import { fetchEvents} from '../../thunks/fetchEvents';
import { fetchLocation } from '../../thunks/fetchLocation';
import { reverseGeocode } from '../../helper';

jest.mock('../../helper')
jest.mock('../../thunks/fetchEvents')
jest.mock('../../thunks/fetchLocation')


describe('BodyForm', () => {

  const mockFetchEvents = jest.fn().mockImplementation(() => Promise.resolve({status: 'ok'}));
  const mockFetchLocation = jest.fn().mockImplementation(() => Promise.resolve({status: 'ok'}));
  const mockSetLocation = jest.fn().mockImplementation(() => Promise.resolve({status: 'ok'}));
  const mockGoBack = jest.fn();
  const mockEvent = {
    preventDefault: jest.fn(),
    target: {id: 'status', value: 'ok, I guess'}
  }

  const mockProps = {
    events: ['one', 'two', 'three'],
    selectedEvent: {eventId: 1},
    location: {lat:1, lng:2},
    fetchEvents: mockFetchEvents,
    fetchLocation: mockFetchLocation,
    setLocation: mockSetLocation,
    history: {goBack: mockGoBack}
  }

  const mockGeolocation = {
    getCurrentPosition: jest.fn().mockImplementation((callback) => {
      return callback({coords: {latitude: 1, longitude:2}})
    })
  }

  let wrapper;

  beforeEach(()=>{
    wrapper = shallow(<BodyForm {...mockProps} />);
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should call handleChange on text input change', () => {
    wrapper = mount(<BodyForm {...mockProps} />);
    const spy = jest.spyOn(wrapper.instance(), 'handleChange');

    wrapper.instance().forceUpdate();
    wrapper.find('#locationName').simulate('change', mockEvent);
    expect(spy).toHaveBeenCalled();
  })

  it('should update state when handleChange is called', () => {
    wrapper.instance().handleChange(mockEvent)
    expect(wrapper.state('status')).toEqual('ok, I guess');
  })

  it('should call handleSubmit on form submission', () => {
    wrapper = mount(<BodyForm {...mockProps} />);
    const spy = jest.spyOn(wrapper.instance(), 'handleSubmit');

    wrapper.instance().forceUpdate();
    wrapper.find('form').simulate('submit', mockEvent);
    expect(spy).toBeCalled();
  })

  it('should call mockFetchLocation when handleSubmit is called', async () => {
    await wrapper.instance().handleSubmit(mockEvent);
    expect(mockFetchLocation).toBeCalled();
  })

  it('should call mockFetchEvents when handleSubmit is called', async () => {
    await wrapper.instance().handleSubmit(mockEvent);
    expect(mockFetchEvents).toBeCalled();
  })

  it('should call useCurrent when checkbox is checked', () => {
    wrapper = mount(<BodyForm {...mockProps} />);
    const spy = jest.spyOn(wrapper.instance(), 'useCurrent');

    wrapper.instance().forceUpdate();
    wrapper.find('#use-current-location').simulate('change', mockEvent);
    expect(spy).toBeCalled();
  })

  it('should update the state when useCurrent is called', () => {
    global.navigator.geolocation = mockGeolocation;
    wrapper.instance().useCurrent(mockEvent);
    expect(wrapper.state('useCurrent')).toEqual(true);
  })

  it('should call mockSetLocation when useCurrent is called', async () => {
    global.navigator.geolocation = mockGeolocation;
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      ok: true,
      json: () => Promise.resolve({results:[{address_components:[{types:['locality'],short_name:'JesusH!'}]}]})
      })
    )

    await wrapper.instance().useCurrent(mockEvent)
    expect(mockSetLocation).toBeCalled()
  })

  it('should return to / when goHome is called', () => {
    wrapper.instance().goHome();
    expect(mockGoBack).toBeCalled();
  })

})

describe('mapStateToProps', () => {
  it('should return an object with a location', () => {
    const mockState = {
      events: ['one', 'two', 'three'],
      randomKey: 'this shouldn\'t show up',
      selectedEvent: {eventId: 1},
      location: {lat: 1, lng: 2}
    }
    const expected = {
      location: {lat: 1, lng: 2}
    }
    const mappedProps = mapStateToProps(mockState);

    expect(mappedProps).toEqual(expected);
  })
})

describe('mapDispatchToProps', () => {

  it('calls dispatch with fetchEvents action when fetchEvents is called', () => {
    const mockDispatch = jest.fn();
    const mockUrl = 'someUrl';
    const actionToDispatch = fetchEvents(mockUrl);
    const mappedProps = mapDispatchToProps(mockDispatch);

    mappedProps.fetchEvents(mockUrl);
    expect(mockDispatch).toBeCalledWith(actionToDispatch)
  })


  it('calls dispatch with fetchLocation action when fetchLocation is called', () => {
    const mockDispatch = jest.fn();
    const mockUrl = 'someUrl';
    const actionToDispatch = fetchLocation(mockUrl);
    const mappedProps = mapDispatchToProps(mockDispatch);

    mappedProps.fetchLocation(mockUrl);
    expect(mockDispatch).toBeCalledWith(actionToDispatch)
  })

    it('calls dispatch with locationFetchSuccess action when setLocation is called', () => {
    const mockDispatch = jest.fn();
    const mockLocation = 'someLocation';
    const actionToDispatch = locationFetchSuccess(mockLocation);
    const mappedProps = mapDispatchToProps(mockDispatch);

    mappedProps.setLocation(mockLocation);
    expect(mockDispatch).toBeCalledWith(actionToDispatch)
  })
})













