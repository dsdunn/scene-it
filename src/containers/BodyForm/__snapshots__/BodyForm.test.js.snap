// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BodyForm should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BodyForm
    events={
        Array [
            "one",
            "two",
            "three",
          ]
    }
    fetchEvents={[Function]}
    fetchLocation={[Function]}
    history={
        Object {
            "goBack": [Function],
          }
    }
    location={
        Object {
            "lat": 1,
            "lng": 2,
          }
    }
    selectedEvent={
        Object {
            "eventId": 1,
          }
    }
    setLocation={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <span
          className="body-form-logo"
          onClick={[Function]}
>
          <a
                    href="#"
          >
                    Scene-It
          </a>
</span>,
        <form
          className="body-form"
          onSubmit={[Function]}
>
          <div
                    className="user-select"
          >
                    <input
                              id="use-current-location"
                              onChange={[Function]}
                              type="checkbox"
                    />
                    <label
                              htmlFor="use-current-location"
                    >
                              Use my current location
                    </label>
          </div>
          <div
                    className="query-info"
          >
                    <input
                              id="locationName"
                              onChange={[Function]}
                              placeholder="location"
                              value=""
                    />
                    <input
                              id="keywords"
                              onChange={[Function]}
                              placeholder="keywords (optional)"
                    />
                    <button>
                              Get Events
                    </button>
          </div>
</form>,
      ],
      "className": "header",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <a
            href="#"
>
            Scene-It
</a>,
          "className": "body-form-logo",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Scene-It",
            "href": "#",
          },
          "ref": null,
          "rendered": "Scene-It",
          "type": "a",
        },
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="user-select"
>
              <input
                            id="use-current-location"
                            onChange={[Function]}
                            type="checkbox"
              />
              <label
                            htmlFor="use-current-location"
              >
                            Use my current location
              </label>
</div>,
            <div
              className="query-info"
>
              <input
                            id="locationName"
                            onChange={[Function]}
                            placeholder="location"
                            value=""
              />
              <input
                            id="keywords"
                            onChange={[Function]}
                            placeholder="keywords (optional)"
              />
              <button>
                            Get Events
              </button>
</div>,
          ],
          "className": "body-form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  id="use-current-location"
                  onChange={[Function]}
                  type="checkbox"
/>,
                <label
                  htmlFor="use-current-location"
>
                  Use my current location
</label>,
              ],
              "className": "user-select",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "id": "use-current-location",
                  "onChange": [Function],
                  "type": "checkbox",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Use my current location",
                  "htmlFor": "use-current-location",
                },
                "ref": null,
                "rendered": "Use my current location",
                "type": "label",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  id="locationName"
                  onChange={[Function]}
                  placeholder="location"
                  value=""
/>,
                <input
                  id="keywords"
                  onChange={[Function]}
                  placeholder="keywords (optional)"
/>,
                <button>
                  Get Events
</button>,
              ],
              "className": "query-info",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "id": "locationName",
                  "onChange": [Function],
                  "placeholder": "location",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "id": "keywords",
                  "onChange": [Function],
                  "placeholder": "keywords (optional)",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Get Events",
                },
                "ref": null,
                "rendered": "Get Events",
                "type": "button",
              },
            ],
            "type": "div",
          },
        ],
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <span
            className="body-form-logo"
            onClick={[Function]}
>
            <a
                        href="#"
            >
                        Scene-It
            </a>
</span>,
          <form
            className="body-form"
            onSubmit={[Function]}
>
            <div
                        className="user-select"
            >
                        <input
                                    id="use-current-location"
                                    onChange={[Function]}
                                    type="checkbox"
                        />
                        <label
                                    htmlFor="use-current-location"
                        >
                                    Use my current location
                        </label>
            </div>
            <div
                        className="query-info"
            >
                        <input
                                    id="locationName"
                                    onChange={[Function]}
                                    placeholder="location"
                                    value=""
                        />
                        <input
                                    id="keywords"
                                    onChange={[Function]}
                                    placeholder="keywords (optional)"
                        />
                        <button>
                                    Get Events
                        </button>
            </div>
</form>,
        ],
        "className": "header",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <a
              href="#"
>
              Scene-It
</a>,
            "className": "body-form-logo",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Scene-It",
              "href": "#",
            },
            "ref": null,
            "rendered": "Scene-It",
            "type": "a",
          },
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="user-select"
>
                <input
                                id="use-current-location"
                                onChange={[Function]}
                                type="checkbox"
                />
                <label
                                htmlFor="use-current-location"
                >
                                Use my current location
                </label>
</div>,
              <div
                className="query-info"
>
                <input
                                id="locationName"
                                onChange={[Function]}
                                placeholder="location"
                                value=""
                />
                <input
                                id="keywords"
                                onChange={[Function]}
                                placeholder="keywords (optional)"
                />
                <button>
                                Get Events
                </button>
</div>,
            ],
            "className": "body-form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <input
                    id="use-current-location"
                    onChange={[Function]}
                    type="checkbox"
/>,
                  <label
                    htmlFor="use-current-location"
>
                    Use my current location
</label>,
                ],
                "className": "user-select",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "id": "use-current-location",
                    "onChange": [Function],
                    "type": "checkbox",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Use my current location",
                    "htmlFor": "use-current-location",
                  },
                  "ref": null,
                  "rendered": "Use my current location",
                  "type": "label",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <input
                    id="locationName"
                    onChange={[Function]}
                    placeholder="location"
                    value=""
/>,
                  <input
                    id="keywords"
                    onChange={[Function]}
                    placeholder="keywords (optional)"
/>,
                  <button>
                    Get Events
</button>,
                ],
                "className": "query-info",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "id": "locationName",
                    "onChange": [Function],
                    "placeholder": "location",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "id": "keywords",
                    "onChange": [Function],
                    "placeholder": "keywords (optional)",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Get Events",
                  },
                  "ref": null,
                  "rendered": "Get Events",
                  "type": "button",
                },
              ],
              "type": "div",
            },
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
